esp-hal/src/gpio/mod.rs

```diff
@@ -883,22 +890,50 @@ impl InterruptConfigurable for Io {
 
 #[ram]
 extern "C" fn user_gpio_interrupt_handler() {
-    USER_INTERRUPT_HANDLER.call();
-
-    default_gpio_interrupt_handler();
+    handle_pin_interrupts(|| USER_INTERRUPT_HANDLER.call());
 }
 
 #[ram]
 extern "C" fn default_gpio_interrupt_handler() {
-    handle_pin_interrupts(on_pin_irq);
+    handle_pin_interrupts(|| ());
 }
 
 #[ram]
-fn on_pin_irq(pin_nr: u8) {
-    // FIXME: async handlers signal completion by disabling the interrupt, but this
-    // conflicts with user handlers.
-    set_int_enable(pin_nr, 0, 0, false);
-    asynch::PIN_WAKERS[pin_nr as usize].wake(); // wake task
+fn handle_pin_interrupts(user_handler: fn()) {
+    let intrs_bank0 = InterruptStatusRegisterAccess::Bank0.interrupt_status_read();
+
+    #[cfg(gpio_bank_1)]
+    let intrs_bank1 = InterruptStatusRegisterAccess::Bank1.interrupt_status_read();
+
+    user_handler();
+
+    // Clear interrupt bits early so that we catch repeat interrupts with larger
+    // probability.
+    Bank0GpioRegisterAccess::write_interrupt_status_clear(intrs_bank0);
+    #[cfg(gpio_bank_1)]
+    Bank1GpioRegisterAccess::write_interrupt_status_clear(intrs_bank1);
+
+    let banks = [
+        (GpioRegisterAccess::Bank0, intrs_bank0),
+        #[cfg(gpio_bank_1)]
+        (GpioRegisterAccess::Bank1, intrs_bank1),
+    ];
+
+    for (bank, intrs) in banks {
+        // Get the mask of active async pins and also unmark them in the same go.
+        let async_pins = bank.async_operations().fetch_and(!intrs, Ordering::Release);
+
+        // Wake up the tasks
+        let mut intr_bits = intrs & async_pins;
+        while intr_bits != 0 {
+            let pin_pos = intr_bits.trailing_zeros();
+            intr_bits -= 1 << pin_pos;
+
+            let pin_nr = pin_pos as u8 + bank.offset();
+            asynch::PIN_WAKERS[pin_nr as usize].wake();
+            set_int_enable(pin_nr, 0, 0, false);
```

Comment:
Could this race against the user calling `listen()` on the other core?

Reply:
If the interrupt handler runs on one core, and while handling you drop the future on the other and start awaiting something again, then maybe. But I think that will just resolve the future immediately on the first poll.

Reply:
I think the interrupt handler would end up unlistening what the user just `listen()`ed.

```Rust
let fut = pin.wait_for_whatever();
fut.poll(...);

// interrupt handler starts running

drop(fut); // user cancels, which unlistens but the handler is already running.

pin.listen(...); // user listens

// handler finishes and calls unlisten.

// user is now unaware that their call to listen is undone.
```

Code looks weird because I'm trying to simulate cancellation.